In [1]:
# Basic Libraries
import numpy as np
import pandas as pd

In [2]:
from sklearn.model_selection import train_test_split

In [4]:
dataa= pd.read_csv('C:/Users/kavya/OneDrive/Desktop/diabetes.csv')

In [5]:
dataa.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [6]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.metrics import accuracy_score

#### With the help of tendorflow we designed a deep learning model that predicts whether a person has diabetes or not by understanding underlying patterns in the factors that contribute to the testing positive of diabetes

### Dividing the data into test and train sets

In [7]:
y= dataa['Outcome']
x=dataa.drop(['Outcome'],axis=1)

In [8]:
xtrain,xtest,ytrain,ytest= train_test_split(x,y,test_size=0.15,stratify=y)
print(xtrain.shape)
print(xtest.shape)
print(ytrain.shape)
print(ytest.shape)

(652, 8)
(116, 8)
(652,)
(116,)


#### Creating DL model

In [17]:
model= Sequential()
model.add(Dense(units=256, activation ='relu', input_shape=([8])))
model.add(Dense(units=156, activation ='relu'))
model.add(Dense(units=1, activation ='sigmoid'))
# we added only two hidden layers to our code
model.summary()
model.compile(loss='binary_crossentropy', optimizer='adam', metrics='accuracy')

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_9 (Dense)             (None, 256)               2304      
                                                                 
 dense_10 (Dense)            (None, 156)               40092     
                                                                 
 dense_11 (Dense)            (None, 1)                 157       
                                                                 
Total params: 42,553
Trainable params: 42,553
Non-trainable params: 0
_________________________________________________________________


#### Fitting the model

In [21]:
num_epochs= 50
model.fit(xtrain,ytrain,epochs=num_epochs, steps_per_epoch=200, validation_data=(xtest ,ytest))

Epoch 1/50
200/200 [==============================] - 0s 2ms/step - loss: 0.3965 - accuracy: 0.7987 - val_loss: 0.6976 - val_accuracy: 0.7586
Epoch 2/50
200/200 [==============================] - 0s 1ms/step - loss: 0.3798 - accuracy: 0.8062 - val_loss: 0.7294 - val_accuracy: 0.7241
Epoch 3/50
200/200 [==============================] - 0s 1ms/step - loss: 0.3812 - accuracy: 0.8163 - val_loss: 0.8009 - val_accuracy: 0.7155
Epoch 4/50
200/200 [==============================] - 0s 1ms/step - loss: 0.3754 - accuracy: 0.8200 - val_loss: 0.6691 - val_accuracy: 0.7069
Epoch 5/50
200/200 [==============================] - 0s 1ms/step - loss: 0.3722 - accuracy: 0.8188 - val_loss: 0.7373 - val_accuracy: 0.7241
Epoch 6/50
200/200 [==============================] - 0s 1ms/step - loss: 0.3839 - accuracy: 0.8213 - val_loss: 0.6736 - val_accuracy: 0.7414
Epoch 7/50
200/200 [==============================] - 0s 1ms/step - loss: 0.3543 - accuracy: 0.8413 - val_loss: 0.7765 - val_accuracy: 0.7155
Epoch 

#### Evaluating

In [31]:
y_pred_test= model.predict(xtest)
y_pred_test= [0 if val<0.5 else 1 for val in y_pred]

4/4 [==============================] - 0s 6ms/step


In [32]:
# checking the accuracy score for the test dataset 
print("Accuracy score (test): " , accuracy_score(ytest,y_pred_test))

Accuracy score (test):  0.7241379310344828


In [33]:
# evaluating the model for train set 
y_pred_train= model.predict(xtrain)
y_pred_train= [0 if val<0.5 else 1 for val in y_pred_train]

21/21 [==============================] - 0s 2ms/step


In [34]:
# checking the accuracy score for the train dataset 
print("Accuracy score (test): " , accuracy_score(ytrain,y_pred_train))

Accuracy score (test):  0.8773006134969326


### Conclusion 

We observed that the accuracy scores of:
-> Train set: 72.41%
-> Test set: 87.7% 
This suggests that our model could generalize well to unseen data, indicating its potential effectiveness in predicting whether a person has diabetes or not. With further improvement, we can use it to predict other diseases. Our initial effort to identify patients at greater risk of developing DKA and hypertension could be enhanced by using this model. Overall, this outcome is promising but it requires careful analysis to increase model efficiency. 



